pkg/webhooks/resource/validation_test.go

```diff
@@ -521,6 +522,380 @@ func TestValidate_failure_action_overrides(t *testing.T) {
+				"check-label-app": "validation error: The label 'app' is required. rule check-label-app failed at path /metadata/labels/",
+			},
+		},
+		{
+			rawPolicy: []byte(`
+				{
+					"apiVersion": "kyverno.io/v1",
+					"kind": "ClusterPolicy",
+					"metadata": {
+					   "name": "check-label-app"
+					},
+					"spec": {
+					   "validationFailureAction": "enforce",
+					   "validationFailureActionOverrides":
+							[
+								{
+									"action": "audit",
+									"namespaces": [
+									  "dev"
+									],
+									"namespaceSelector": {
+										"matchExpressions": [{
+										  "key" : "kubernetes.io/metadata.name",
+                      "operator": "In",
+                      "values": [
+										 	  "prod"
+											]
+										}]
+									}
+								}
+							],
+					   "rules": [
+						  {
+							 "name": "check-label-app",
+							 "match": {
+								"resources": {
+								   "kinds": [
+									  "Pod"
+								   ]
+								}
+							 },
+							 "validate": {
+								"message": "The label 'app' is required.",
+								"pattern": {
+									"metadata": {
+										"labels": {
+											"app": "?*"
+										}
+									}
+								}
+							}
+						  }
+					   ]
+					}
+			 	}
+			`),
+			rawResource: []byte(`
+				{
+					"apiVersion": "v1",
+					"kind": "Pod",
+					"metadata": {
+					   "name": "test-pod",
+						 "namespace": "dev"
+					},
+					"spec": {
+					   "containers": [
+						  {
+							 "name": "nginx",
+							 "image": "nginx:latest"
+						  }
+					   ]
+					}
+				 }
+			`),
+			blocked: false,
+			rawResourceNamespaceLabels: map[string]string{
+				"kubernetes.io/metadata.name": "dev",
+			},
+		},
+		{
+			rawPolicy: []byte(`
+				{
+					"apiVersion": "kyverno.io/v1",
+					"kind": "ClusterPolicy",
+					"metadata": {
+					   "name": "check-label-app"
+					},
+					"spec": {
+					   "validationFailureAction": "enforce",
+					   "validationFailureActionOverrides":
+							[
+								{
+									"action": "audit",
+									"namespaces": [
+									  "dev"
+									],
+									"namespaceSelector": {
+										"matchExpressions": [{
+										  "key" : "kubernetes.io/metadata.name",
+                      "operator": "In",
+                      "values": [
+										 	  "prod"
+											]
+										}]
+									}
+								}
+							],
+					   "rules": [
+						  {
+							 "name": "check-label-app",
+							 "match": {
+								"resources": {
+								   "kinds": [
+									  "Pod"
+								   ]
+								}
+							 },
+							 "validate": {
+								"message": "The label 'app' is required.",
+								"pattern": {
+									"metadata": {
+										"labels": {
+											"app": "?*"
+										}
+									}
+								}
+							}
+						  }
+					   ]
+					}
+			 	}
+			`),
+			rawResource: []byte(`
+				{
+					"apiVersion": "v1",
+					"kind": "Pod",
+					"metadata": {
+					   "name": "test-pod",
+						 "namespace": "prod"
```

Comment:
Does it require all `namespaces` and `namespaceSelector` match to overwrite the validation action? The policy specifies the namespace`dev` while the resource is in `prod`, shouldn't it be blocked?

Reply:
Either the ns specified by `namespaces` or `namespaceSelector` would be overwritten. So they are independent.

In this case, as `prod` is specified in the `namespaceSelector` and the action is `audit`, so it wouldn't be blocked.

Reply:
Is it a logical OR when evaluating `namespaces` and `namespaceSelector`?

Reply:
Yes, exactly.
If this is not what we assume, I can change that.

Reply:
Yeah we probably want to change it to logical AND to make it consistent with these two fields in the match block.

Reply:
Sounds good.
Also added test cases for `validationFailureAction=audit` and overrides the action to enforce.